csmith-fuzzing/predicate.py

```diff
@@ -0,0 +1,236 @@
+reducing.add_argument(
+    "--expect-compile-fail",
+    action="store_true",
+    help="Exit non-zero if `rustc` successfully compiles the emitted bindings.")
+reducing.add_argument(
+    "--rustc-grep",
+    type=str,
+    help="Exit non-zero if the output from compiling the bindings with `rustc` does not contain the given regexp pattern")
+
+reducing.add_argument(
+    "--no-layout-tests",
+    action="store_false",
+    dest="layout_tests",
+    help="Do not run the compiled bindings' layout tests.")
+reducing.add_argument(
+    "--expect-layout-tests-fail",
+    action="store_true",
+    help="Exit non-zero if the compiled bindings' layout tests pass.")
+reducing.add_argument(
+    "--layout-tests-grep",
+    type=str,
+    help="Exit non-zero if the output of running the compiled bindings' layout tests does not contain the given regexp pattern.")
+
+################################################################################
+
+class ExitOne(Exception):
+    pass
+
+def exit_1(msg, child=None):
+    print(msg)
+
+    if child:
+        print("---------- stdout ----------------------------------------------")
+        print(decode(child.stdout))
+        print("---------- stderr ----------------------------------------------")
+        print(decode(child.stderr))
+
+    raise ExitOne()
+
+def main():
+    args = parser.parse_args()
+    os.environ["RUST_BACKTRACE"] = "1"
+
+    exit_code = 0
+    try:
+        bindings = new_temp_file(prefix="bindings-", suffix=".rs")
+        run_bindgen(args, bindings)
+
+        if args.rustc:
+            test_exe = new_temp_file(prefix="layout-tests-")
+            run_rustc(args, bindings, test_exe)
+
+            if args.layout_tests:
+                run_layout_tests(args, test_exe)
+    except ExitOne:
+        exit_code = 1
+    except Exception as e:
+        exit_code = 2
+        print("Unexpected exception:", e)
+
+    for p in TEMP_FILES:
+        try:
+            os.remove(path)
+        except:
+            pass
+
+    sys.exit(exit_code)
+
+def run(cmd, **kwargs):
+    print("Running:", cmd)
+    return subprocess.run(cmd, **kwargs)
+
+def decode(f):
+    return f.decode(encoding="utf-8", errors="ignore")
+
+TEMP_FILES = []
+
+def new_temp_file(prefix, suffix=None):
+    temp = NamedTemporaryFile(delete=False, prefix=prefix, suffix=suffix)
+    temp.close()
+    TEMP_FILES.append(temp.name)
+    return temp.name
+
+def contains(pattern, lines):
+    for line in lines:
+        if re.match(pattern, line):
+            return True
+    return False
+
+def regexp(pattern):
+    if not pattern.startswith("^"):
+        pattern = ".*" + pattern
+    if not pattern.endswith("$"):
+        pattern = pattern + ".*"
+    return re.compile(pattern)
+
+def run_bindgen(args, bindings):
+    child = run(
+        ["cargo", "run", "--", args.input, "-o", bindings] + shlex.split(args.bindgen_args),
+        cwd=os.path.abspath(os.path.dirname(sys.argv[0])),
```

Comment:
I don't quite get why do we need this.

If I run this script like this:

```shell
~/rust-bindgen $ ./csmith-fuzzing/predicate.py csmith-fuzzing/csmith.h
```

It wouldn't find `csmith.h`, because real path would be `./csmith-fuzzing/csmith-fuzzing/csmith.h`.

If we would run:

```shell
~/rust-bindgen/csmith-fuzzing $ ./predicate.py csmith.h
```
we would be already in `csmith-fuzzing`.

Maybe it would be better to drop `cwd` and do: 

```shell
cargo run --manifest-path $(basedir)/../Cargo.toml ...
``` 
?

Reply:
*Great* catch!